substrate/client/db/src/cache.rs

```diff
@@ -0,0 +1,412 @@
+/// previous entry in the cache, allowing us to traverse back in time in list-style.
+#[cfg_attr(test, derive(Debug, PartialEq))]
+struct StorageEntry<N, T> {
+	/// None if valid from the beginning
+	prev_valid_from: Option<N>,
+	/// None means that we do not know the value starting from `valid_from` block
+	value: Option<T>,
+}
+
+impl<Block, T> DbCacheList<Block, T>
+	where
+		Block: BlockT,
+		<<Block as BlockT>::Header as HeaderT>::Number: As<u64>,
+		T: Clone + PartialEq + Codec,
+{
+	/// Creates new cache list.
+	fn new(db: Arc<KeyValueDB>, meta_key: &'static [u8], column: Option<u32>) -> ClientResult<Self> {
+		let best_entry = RwLock::new(db.get(COLUMN_META, meta_key)
+			.map_err(db_err)
+			.and_then(|block| match block {
+				Some(block) => {
+					let valid_from = db_key_to_number(&block)?;
+					read_storage_entry::<Block, T>(&*db, column, valid_from)
+						.map(|entry| Some(Entry {
+							valid_from,
+							value: entry
+								.expect("meta entry references the entry at the block; storage entry at block exists when referenced; qed")
+								.value,
+						}))
+				},
+				None => Ok(None),
+			})?);
+
+		Ok(DbCacheList {
+			db,
+			column,
+			meta_key,
+			best_entry,
+		})
+	}
+
+	/// Gets the best known entry.
+	pub fn best_entry(&self) -> Option<Entry<<<Block as BlockT>::Header as HeaderT>::Number, T>> {
+		self.best_entry.read().clone()
+	}
+
+	/// Commits the new best pending value to the database. Returns Some if best entry must
+	/// be updated after transaction is committed.
+	pub fn commit_best_entry(&self, transaction: &mut DBTransaction, valid_from: <<Block as BlockT>::Header as HeaderT>::Number, pending_value: Option<T>) -> Option<Entry<<<Block as BlockT>::Header as HeaderT>::Number, T>> {
+		let best_entry = self.best_entry();
+		let update_best_entry = match (best_entry.as_ref().and_then(|a| a.value.as_ref()), pending_value.as_ref()) {
+			(Some(best_value), Some(pending_value)) => best_value != pending_value,
+			(None, Some(_)) | (Some(_), None) => true,
+			(None, None) => false,
+		};
+		if !update_best_entry {
+			return None;
+		}
+
+		let valid_from_key = number_to_db_key(valid_from);
+		transaction.put(COLUMN_META, self.meta_key, &valid_from_key);
+		transaction.put(self.column, &valid_from_key, &StorageEntry {
+			prev_valid_from: best_entry.map(|b| b.valid_from),
+			value: pending_value.clone(),
+		}.encode());
+
+		Some(Entry {
+			valid_from,
+			value: pending_value,
+		})
+	}
+
+	/// Updates the best in-memory cache entry. Must be called after transaction with changes
+	/// from commit_best_entry has been committed.
+	pub fn update_best_entry(&self, best_entry: Option<Entry<<<Block as BlockT>::Header as HeaderT>::Number, T>>) {
+		*self.best_entry.write() = best_entry;
+	}
+
+	/// Prune all entries from the beginning up to the block (including entry at the number). Returns
+	/// the number of pruned entries. Pruning never deletes the latest entry in the cache.
```

Comment:
Well, it should actually prune all but the last one. Like if it is called so that it should prune entries for blocks `[10, 20, 30, 40]` and the entry#40 is the last entry, entries `[10, 20, 30]` should be removed. And if there's entry for block#50, entries for `[10, 20, 30, 40]` are removed.

If you believe it doesn't work that way, consider this an error.

Reply:
Do you have test for that? Because if I understand [L218](https://github.com/paritytech/polkadot/pull/251/files/ed1cbfd2d77c41be8a8924701e429a32a602fb84..1f57347801708087be0c2e97eb743defb2e91536#diff-d76f82a0a8543731181e31ea8f469e5eR218) correctly, we _early quit_ the function without actually pruning if our loop finds the last entry, no?

Reply:
@svyatonik was this point resolved?